include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,486 @@
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param buf The data buffer to receive sensor data sample.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_read_sample(
+		senss_sensor_handle_t handle,
+		void *buf, int size);
+
+/**
+ * @brief Trigger a sensor to do self calibration
+ *
+ * Some sensor support auto self calibration.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+int senss_self_calibration(
+		senss_sensor_handle_t handle);
+
+/**
+ * @brief Sensor data event receive callback.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param buf The data buffer with sensor data.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @param param The user private parameter.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+typedef int (*senss_data_event_t)(
+		senss_sensor_handle_t handle,
+		void *buf, int size, void *param);
+
+/**
+ * @brief Register sensor data event receive callback.
+ *
+ * User can use this API to register a data event receive callback function.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param callback The data event receive callback function.
+ *
+ * @param param The user private parameter.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_register_data_event_callback(
+		senss_sensor_handle_t handle,
+		senss_data_event_t callback, void *param);
+
+/**
+ * @brief Set batching buffer
+ *
+ * Application can allocated a batching buffer and set to sensor
+ * subsystem runtime.
+ *
+ * @param buf The buffer pointer to be set.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_batching_set_buffer(
+		void *buf, int size);
+
+/**
+ * @brief Set report latency of a sensor instance.
+ *
+ * Report latency controls how long the sensor data events of the sensor instance will be buffered
+ * in the batching buffer before deliver to its client.
+ *
+ * For 0 latency value, it's means sensor data event will not buffered, and will report to its client
+ * immediately.
+ *
+ * The maximum sensor data events of this sensor instance is calculated as: latency /
+ * report interval.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param latency The latency in micro seconds.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_set_report_latency(
+		senss_sensor_handle_t handle, uint64_t latency);
+
+/**
+ * @brief Get current report latency value of a sensor instance.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param latency The value buffer to receive report latency in micro seconds.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_get_report_latency(
+		senss_sensor_handle_t handle, uint64_t *latency);
```

Comment:
In `senss` design, FIFO is specific for sensor HW FIFO, handled by physical sensors with HW FIFO supported sensor chip, its transparent to application. 

Here the `latency` only impacts the batching, which management by `senss`, since `senss` is single thread scheduling, and `latency` only tells `senss` how much samples (batching size = report latency / report interval) should be buffered before report to application clients, not impact internal sensor scheduling.

So, I think there should be no tricky logic and race conditions. The things need consider is when buffer size changed due to `latency` or `interval` settings, flush existing samples in batch buffer, or just apply new buffer size and flush on next time ?

Reply:
I don't think you can separate interval from latency, this is also true when performing arbitration. Imagine the arbitration process. It would need to iterate through the client configuration requests. These requests are more complex than just taking the `min` latency and `max` interval. The reason is that you'll need to keep the buffers data around depending on the range of latency settings. 

1. Client 1 sets the latency to 200ms
2. Client 2 sets the latency to 500ms
3. When data is there and 200ms pass, we can send the buffer to client 1, but the buffer can't be recycled yet until client 2 gets its data 300ms later. 

Overall, since these values don't really make sense apart, I don't see a need to split them. It makes the system vulnerable to mistakes.

Reply:
I think the purpose of batching and buffer mode is to save more system power since we can let sensor hub batch and buffer sensor data when host is in sleep mode. 

 Since all sensors shared single batch buffer, we can make things more simple, if a sensor's latency is reached and need to flush, such as the example in # 1, we can flush all existing sensor data in the buffer to their clients together, because the host is already waked up due to the latency reached sensor, we can take this chance to flush all existing sensor data in the buffer, then the host will enter sleep again, this strategy can reduce the host wakeup frequency and save more power. For your client 1 (200ms latency) and client 2 (500ms latency) example, with this strategy, host only need wakeup 5 times during 1 seconds, but need wakeup 7 times (5 times for client 1, 2 times for client 2) for separated flush for each clients